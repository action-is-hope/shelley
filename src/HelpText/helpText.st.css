@namespace "HelpText";

.root {
  -st-states: isDisabled, isErrorMessage;
}

.helpText {}
