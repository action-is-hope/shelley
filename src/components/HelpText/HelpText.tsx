/** HelpText.tsx */
import type React from "react";
import { ReactNode, HTMLAttributes, forwardRef } from "react";
import type { Validation } from "../../typings/shared-types";
import Warning from "../icons/Warning";
import { st, classes } from "./helpText.st.css";

interface HelpTextProps extends Validation, React.HTMLProps<HTMLDivElement> {
  /** Whether the description is displayed with lighter text. */
  isDisabled?: boolean;
  /** Whether an error icon is rendered. */
  showErrorIcon?: boolean;
  /** A description for the field. Provides a hint such as specific requirements for what to choose. */
  description?: ReactNode;
  /** An error message for the field. */
  errorMessage?: ReactNode;
  /** Props for the help text description element. */
  descriptionProps?: HTMLAttributes<HTMLElement>;
  /** Props for the help text error message element. */
  errorMessageProps?: HTMLAttributes<HTMLElement>;
  /** Add predefined data-id to ease testing or analytics. */
  includeDataIds?: boolean;
}

function HelpText(props: HelpTextProps, ref?: React.Ref<HTMLDivElement>) {
  const {
    description,
    errorMessage,
    validationState,
    isDisabled,
    showErrorIcon,
    descriptionProps,
    errorMessageProps,
    className: classNameProp,
    includeDataIds = false,
  } = props;
  const isErrorMessage =
    (errorMessage && validationState === "invalid") || false;
  return (
    <>
      {(description || isErrorMessage) && (
        <div
          className={st(
            classes.root,
            { isErrorMessage, isDisabled },
            classNameProp
          )}
          ref={ref}
        >
          {isErrorMessage ? (
            <>
              {showErrorIcon && <Warning />}
              <div
                {...errorMessageProps}
                className={st(classes.helpText)}
                data-id={includeDataIds ? "help--error" : undefined}
              >
                {errorMessage}
              </div>
            </>
          ) : (
            <div
              {...descriptionProps}
              className={st(classes.helpText)}
              data-id={includeDataIds ? "help--description" : undefined}
            >
              {description}
            </div>
          )}
        </div>
      )}
    </>
  );
}

const _HelpText = forwardRef(HelpText);
export { _HelpText as HelpText };
