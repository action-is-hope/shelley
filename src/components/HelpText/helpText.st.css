/* helpText.st.css */
@namespace "HelpText";

.root {
  -st-states: isDisabled, isErrorMessage;
}

.helpText {}