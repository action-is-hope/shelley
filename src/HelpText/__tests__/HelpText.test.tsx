import renderer from "react-test-renderer";
import { HelpText } from "../HelpText";

const description = "Help text";
const errorMessage = "Error message";

describe("HelpText", () => {
  it("renders help text", () => {
    const tree = renderer
      .create(
        <HelpText
          className="custom-class"
          data-testid="help-text-id"
          description={description}
          errorMessage={errorMessage}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
