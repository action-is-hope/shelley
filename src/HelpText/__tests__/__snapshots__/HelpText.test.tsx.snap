// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HelpText renders help text 1`] = `
<div
  className="helpText__Test__root custom-class"
>
  <div
    className="helpText__Test__helpText"
  >
    Help text
  </div>
</div>
`;
